### PR TITLE
fix: catch `Ocaml_typing.Magic_numbers.Cmi.Error(_)` instead of silently failing

### LINKED ISSUE #567: Show friendlier message for mismatching merlin compiler and project compiler versions
If the ocaml compiler version in merlin mismatches the compiler version the project is built with, a user should get exceptions. We should handle those nicely and show an error message (making sure we show the message once per ocaml-lsp instantiation) 

```json
[Trace - 10:19:32 PM] Received response 'textDocument/hover - (15)' in 4ms. Request failed: uncaught exception (-32603).
Error data: {
    "exn": "Magic_numbers.Cmi.Error(_)",
    "backtrace": "Raised at Cmi_format.read_cmi in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/cmi_format.ml\", line 75, characters 6-21\nCalled from Merlin_utils__File_cache.Make.read in file \"ocaml-lsp-server/vendor/merlin/src/utils/file_cache.ml\", line 57, characters 17-36\nRe-raised at Merlin_utils__File_cache.Make.read in file \"ocaml-lsp-server/vendor/merlin/src/utils/file_cache.ml\", line 64, characters 6-15\nCalled from Persistent_env.Persistent_signature.load.(fun) in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/persistent_env.ml\", line 43, characters 18-41\nCalled from Persistent_env.find_pers_struct in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/persistent_env.ml\", line 256, characters 16-58\nCalled from Persistent_env.find in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/persistent_env.ml\", line 303, characters 6-45\nCalled from Env.lookup_ident_module in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2698, characters 16-31\nCalled from Env.lookup_module_components in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2779, characters 23-71\nCalled from Env.lookup_structure_components in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2791, characters 20-70\nCalled from Env.lookup_dot_module in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2880, characters 17-68\nCalled from Env.lookup_module in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2871, characters 23-66\nCalled from Merlin_analysis__Locate.Env_lookup.in_namespaces.(fun) in file \"ocaml-lsp-server/vendor/merlin/src/analysis/locate.ml\", line 671, characters 27-60\nCalled from Stdlib__List.iter in file \"list.ml\", line 110, characters 12-15\nCalled from Merlin_analysis__Locate.Env_lookup.in_namespaces in file \"ocaml-lsp-server/vendor/merlin/src/analysis/locate.ml\", line 654, characters 6-1023\nCalled from Merlin_analysis__Locate.from_longident in file \"ocaml-lsp-server/vendor/merlin/src/analysis/locate.ml\", line 732, characters 8-46\nCalled from Merlin_analysis__Locate.get_doc.(fun) in file \"ocaml-lsp-server/vendor/merlin/src/analysis/locate.ml\", line 832, characters 8-64\nCalled from Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 685, characters 8-12\nRe-raised at Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 687, characters 30-39\nCalled from Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 685, characters 8-12\nRe-raised at Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 687, characters 30-39\nCalled from Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 685, characters 8-12\nRe-raised at Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 687, characters 30-39\nCalled from Merlin_utils__Misc.try_finally in file \"ocaml-lsp-server/vendor/merlin/src/utils/misc.ml\", line 45, characters 8-15\nRe-raised at Merlin_utils__Misc.try_finally in file \"ocaml-lsp-server/vendor/merlin/src/utils/misc.ml\", line 62, characters 10-24\nCalled from Stdlib__Fun.protect in file \"fun.ml\", line 33, characters 8-15\nRe-raised at Stdlib__Fun.protect in file \"fun.ml\", line 38, characters 6-52\nCalled from Mocaml.with_state in file \"ocaml-lsp-server/vendor/merlin/src/kernel/mocaml.ml\", line 18, characters 8-38\nRe-raised at Mocaml.with_state in file \"ocaml-lsp-server/vendor/merlin/src/kernel/mocaml.ml\", line 20, characters 42-53\nCalled from Stdune__Exn_with_backtrace.try_with in file \"submodules/dune/otherlibs/stdune-unstable/exn_with_backtrace.ml\", line 9, characters 8-12\nRe-raised at Stdune__Exn.raise_with_backtrace in file \"submodules/dune/otherlibs/stdune-unstable/exn.ml\", line 36, characters 27-56\nCalled from Fiber.O.(>>|).(fun) in file \"submodules/dune/src/fiber/fiber.ml\", line 288, characters 36-41\nCalled from Fiber.Execution_context.run_jobs in file \"submodules/dune/src/fiber/fiber.ml\", line 204, characters 8-13\n"
}


[Error - 10:19:32 PM] Request textDocument/hover failed.
  Message: uncaught exception
  Code: -32603 
[object Object]
```

```json
[Trace - 10:21:13 PM] Received response 'workspace/symbol - (28)' in 21ms. Request failed: uncaught exception (-32603).
Error data: {
    "exn": "Magic_numbers.Cmi.Error(_)",
    "backtrace": "Raised at Cmt_format.read.(fun) in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/cmt_format.ml\", line 135, characters 11-69\nCalled from Merlin_utils__Misc.try_finally in file \"ocaml-lsp-server/vendor/merlin/src/utils/misc.ml\", line 45, characters 8-15\nRe-raised at Merlin_utils__Misc.try_finally in file \"ocaml-lsp-server/vendor/merlin/src/utils/misc.ml\", line 62, characters 10-24\nCalled from Cmt_format.read_cmt in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/cmt_format.ml\", line 141, characters 8-21\nCalled from Ocaml_lsp_server__Workspace_symbol.symbols_from_cm_file in file \"ocaml-lsp-server/src/workspace_symbol.ml\", line 221, characters 4-32\nCalled from Stdune__List.concat_map.aux in file \"submodules/dune/otherlibs/stdune-unstable/list.ml\", line 44, characters 15-18\nCalled from Ocaml_lsp_server__Workspace_symbol.run.(fun) in file \"ocaml-lsp-server/src/workspace_symbol.ml\", line 291, characters 9-72\nCalled from Stdlib__List.rev_map.rmap_f in file \"list.ml\", line 103, characters 22-25\nCalled from Stdune__List.map in file \"submodules/dune/otherlibs/stdune-unstable/list.ml\", line 5, characters 19-33\nCalled from Stdune__Exn_with_backtrace.try_with in file \"submodules/dune/otherlibs/stdune-unstable/exn_with_backtrace.ml\", line 9, characters 8-12\nRe-raised at Stdune__Exn.raise_with_backtrace in file \"submodules/dune/otherlibs/stdune-unstable/exn.ml\", line 36, characters 27-56\nCalled from Fiber.O.(>>|).(fun) in file \"submodules/dune/src/fiber/fiber.ml\", line 288, characters 36-41\nCalled from Fiber.Execution_context.run_jobs in file \"submodules/dune/src/fiber/fiber.ml\", line 204, characters 8-13\n"
}


[Error - 10:21:13 PM] Request workspace/symbol failed.
  Message: uncaught exception
  Code: -32603 
[object Object]
```

### PR DESCRIPTION
**Pull request by ulugbekna**
Fixes #567 

@rgrinberg Hi, I'm layering another `on_request` on top of `on_request` handling all language requests but somehow the exception is still uncaught. I suspect that I'm misusing fiber context, where the exception is thrown in the scheduler and so not caught. What do you think? Thanks!

### CODE REVIEW

File: ocaml-lsp-server/src/ocaml_lsp_server.ml [hunk 1]
```diff
@@ -781,6 +781,48 @@ let document_symbol (state : State.t) uri =
   in
   Some symbols
 
+let doc_in_request (type resp) (client_req : resp Client_request.t) =
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/ocaml_lsp_server.ml:784]: Client_request.text_document already does this.

**Comment by rgrinberg** [comment]
I think your code is working as intended but it's not solving the problem because the error occurs when we generate the push the diagnostics to the client rather than as a result of any request.

It should be enough to handle this exactly the way we handle meriln-extend issues. The error is displayed in the diagnostics with a clear message and if the user keeps sending requests that are bound to fail, then it's fine. I don't think anyone would be surprised by those failures.

**Comment by ulugbekna** [comment]
> I think your code is working as intended but it's not solving the problem because the error occurs when we generate the push the diagnostics to the client rather than as a result of any request.

Hmm, either I'm missing something or I don't agree. This exception occurs as a result of ocamllsp trying to respond to a certain request because we get this in output:

```
[Trace - 10:19:32 PM] Received response 'textDocument/hover - (15)' in 4ms. Request failed: uncaught exception (-32603).
Error data: {
    "exn": "Magic_numbers.Cmi.Error(_)",
    "backtrace": "Raised at Cmi_format.read_cmi in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/cmi_format.ml\", line 75, characters 6-21\nCalled from Merlin_utils__File_cache.Make.read in file \"ocaml-lsp-server/vendor/merlin/src/utils/file_cache.ml\", line 57, characters 17-36\nRe-raised at Merlin_utils__File_cache.Make.read in file \"ocaml-lsp-server/vendor/merlin/src/utils/file_cache.ml\", line 64, characters 6-15\nCalled from Persistent_env.Persistent_signature.load.(fun) in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/persistent_env.ml\", line 43, characters 18-41\nCalled from Persistent_env.find_pers_struct in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/persistent_env.ml\", line 256, characters 16-58\nCalled from Persistent_env.find in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/persistent_env.ml\", line 303, characters 6-45\nCalled from Env.lookup_ident_module in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2698, characters 16-31\nCalled from Env.lookup_module_components in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2779, characters 23-71\nCalled from Env.lookup_structure_components in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2791, characters 20-70\nCalled from Env.lookup_dot_module in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2880, characters 17-68\nCalled from Env.lookup_module in file \"ocaml-lsp-server/vendor/merlin/src/ocaml/typing/env.ml\", line 2871, characters 23-66\nCalled from Merlin_analysis__Locate.Env_lookup.in_namespaces.(fun) in file \"ocaml-lsp-server/vendor/merlin/src/analysis/locate.ml\", line 671, characters 27-60\nCalled from Stdlib__List.iter in file \"list.ml\", line 110, characters 12-15\nCalled from Merlin_analysis__Locate.Env_lookup.in_namespaces in file \"ocaml-lsp-server/vendor/merlin/src/analysis/locate.ml\", line 654, characters 6-1023\nCalled from Merlin_analysis__Locate.from_longident in file \"ocaml-lsp-server/vendor/merlin/src/analysis/locate.ml\", line 732, characters 8-46\nCalled from Merlin_analysis__Locate.get_doc.(fun) in file \"ocaml-lsp-server/vendor/merlin/src/analysis/locate.ml\", line 832, characters 8-64\nCalled from Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 685, characters 8-12\nRe-raised at Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 687, characters 30-39\nCalled from Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 685, characters 8-12\nRe-raised at Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 687, characters 30-39\nCalled from Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 685, characters 8-12\nRe-raised at Merlin_utils__Std.let_ref in file \"ocaml-lsp-server/vendor/merlin/src/utils/std.ml\", line 687, characters 30-39\nCalled from Merlin_utils__Misc.try_finally in file \"ocaml-lsp-server/vendor/merlin/src/utils/misc.ml\", line 45, characters 8-15\nRe-raised at Merlin_utils__Misc.try_finally in file \"ocaml-lsp-server/vendor/merlin/src/utils/misc.ml\", line 62, characters 10-24\nCalled from Stdlib__Fun.protect in file \"fun.ml\", line 33, characters 8-15\nRe-raised at Stdlib__Fun.protect in file \"fun.ml\", line 38, characters 6-52\nCalled from Mocaml.with_state in file \"ocaml-lsp-server/vendor/merlin/src/kernel/mocaml.ml\", line 18, characters 8-38\nRe-raised at Mocaml.with_state in file \"ocaml-lsp-server/vendor/merlin/src/kernel/mocaml.ml\", line 20, characters 42-53\nCalled from Stdune__Exn_with_backtrace.try_with in file \"submodules/dune/otherlibs/stdune-unstable/exn_with_backtrace.ml\", line 9, characters 8-12\nRe-raised at Stdune__Exn.raise_with_backtrace in file \"submodules/dune/otherlibs/stdune-unstable/exn.ml\", line 36, characters 27-56\nCalled from Fiber.O.(>>|).(fun) in file \"submodules/dune/src/fiber/fiber.ml\", line 288, characters 36-41\nCalled from Fiber.Execution_context.run_jobs in file \"submodules/dune/src/fiber/fiber.ml\", line 204, characters 8-13\n"
}


[Error - 10:19:32 PM] Request textDocument/hover failed.
  Message: uncaught exception
  Code: -32603 
[object Object]
```

Before sending the patch, I looked at in what ways merlin exposes this exception. I think there are two:
- When reporting diagnostics, merlin puts this exception in typer errors (ie it removes all typer errors and reports only this one) -- that's why ocamllsp shows diagnostics on the first line of the file with its own message about mismatching compiler versions
- The exception is thrown on certain command calls, which I'm trying to patch here

**Comment by rgrinberg** [comment]
Okay, then I think I misunderstood the original issue. What's the problem with commands failing if we are reporting the issue in the diagnostics? Is the ugly message in the logs?

**Comment by ulugbekna** [comment]
> Okay, then I think I misunderstood the original issue. What's the problem with commands failing if we are reporting the issue in the diagnostics? Is the ugly message in the logs?

I would assume users do not necessarily understand why a certain doesn't work based on a single diagnostic. If ocamllsp crashed with a message about the cmi mismatch, it would be clear. As of now, it's not, in my opinion. Showing an explicit message is better, and hence the PR. 
---

Regarding the diagnostic, the message mentions "Merlin" in the message. I think we could paraphrase the message to use ocamllsp and give hints how to fix the issue. Is using a regular expression library fine for changing the message? I would go through all diagnostics and apply regex only if the kind is a typer error and if there's only one typer error.

**Comment by rgrinberg** [comment]
> I would assume users do not necessarily understand why a certain doesn't work based on a single diagnostic. If ocamllsp crashed with a message about the cmi mismatch, it would be clear. As of now, it's not, in my opinion. Showing an explicit message is better, and hence the PR.

It seems a bit intrusive to me. For example, when typing, the client requests completion results. It would be annoying to see this message pop up all of a sudden. I would limit this behavior to requests that the user triggers explicitly, like hover for example.

> Regarding the diagnostic, the message mentions "Merlin" in the message. I think we could paraphrase the message to use ocamllsp and give hints how to fix the issue. Is using a regular expression library fine for changing the message? I would go through all diagnostics and apply regex only if the kind is a typer error and if there's only one typer error.

No regex please. Just apply a patch to our merlin fork to make this issue easy to handle.

**Comment by ulugbekna** [comment]
I created https://github.com/rgrinberg/merlin/pull/16 to 
- change the message to mention OCaml LSP rather than Merlin
- mention that certain functionality will not work, so we don't necessarily need to worry much

**Pull request closed without merging**